Add error boundary to dashcards and visualizations

iethree opened the pull request:
## Description

- Extracts our top-level ErrorBoundary component to be more reusable
- Adds some basic default behavior to the ErrorBoundary component to make it easier to use
- Allows the default error component to be overidden for each error boundary
- Adds an ErrorBoundary to the dashcard  and visualizations components so that one erroring visualization won't crash a whole dashboard or the query builder

Before | After
--- | ---
one broken visualization breaks the whole app | one broken visualization breaks just that visualization!
![Screen Shot 2023-01-27 at 3 46 10 PM](https://user-images.githubusercontent.com/30528226/215220696-756d2ff8-d78f-460d-9b4f-ea45827e560f.png) |  ![Screen Shot 2023-01-30 at 1 16 21 PM](https://user-images.githubusercontent.com/30528226/215587026-72a64f27-6229-41a3-ad13-c66a63164c2a.png)
![Screen Shot 2023-01-30 at 1 25 27 PM](https://user-images.githubusercontent.com/30528226/215587306-961fd087-4d63-47dd-a825-b668266a8fae.png) | ![Screen Shot 2023-01-30 at 1 15 11 PM](https://user-images.githubusercontent.com/30528226/215587143-7a01c590-a548-40a1-8ad9-3a27f992ab52.png)
![Screen Shot 2023-01-30 at 1 29 28 PM](https://user-images.githubusercontent.com/30528226/215588095-846079b6-3b72-42d3-9b75-504399281037.png) | ![Screen Shot 2023-01-30 at 1 28 46 PM](https://user-images.githubusercontent.com/30528226/215588100-bf1da504-371b-4c5e-abbb-d65f8d50ecad.png)






## How to Test

- add a couple of questions to a dashboard with different visualizations
- pick your favorite visualization (we know it's LineAreaBarChart 😉 ), and add a `throw new Error('foo')` to the render() method.
- see that instead of the whole dashboard crashing, you get a little error message for just the broken visualization.
- open the query builder and select your broken visualization type 
- see that only the viz preview breaks! not the whole query builder!
- open the collections view and pin a broken viz
- see that only the viz preview for that item breaks! not the whole collection browser!

iethree committed:
add error boundary to dashcards

iethree commented on frontend/src/metabase/containers/ErrorPages.jsx:
```diff
@@ -54,3 +55,14 @@ export const Archived = ({ entityName, linkTo }) => (
   </ErrorPageRoot>
 );
+
+export const SmallGenericError = ({
```
Added a smaller generic error component because the other one was designed to be a whole page, and was too large for many individual components.

iethree committed:
fix type error

alxnddr commented on frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx:
```diff
@@ -311,46 +313,48 @@ function DashCard({
-        isSlow={isSlow}
-        isPreviewing={isPreviewingCard}
-        isFullscreen={isFullscreen}
+    <ErrorBoundary>
```
That's a great improvement! I believe it would be even better to put it around in `Visualization` component to prevent page crashes in collections view, query builder, and publicly shared questions

iethree replied:
![why-not-both](https://user-images.githubusercontent.com/30528226/215515746-6243e4bb-d5d6-49a7-965e-e6330177e5e6.jpg)

I think there's some logic in dashcard that we want to catch errors on as well, but the beauty of error boundaries is we can wrap things at multiple levels.

ranquild reviewed (commented):
WDYT about styling this error to look similar to the permissions error? 
<img width="1105" alt="Screenshot 2023-01-30 at 14 49 17" src="https://user-images.githubusercontent.com/8542534/215481651-5571d178-5c4a-4ba6-b7fd-2d62716b27c4.png">

iethree commented:
> WDYT about styling this error to look similar to the permissions error? 

The design challenge is that it needs to fit in a potentially very small space. At first I tried the generic error page message and it was hilariously bad. I'll ask the design team for help

iethree committed:
update error catching behavior and allow override component

iethree committed:
update default error component

iethree committed:
add error boundary to visualization component

iethree commented:
Made a few updates
1. Added a [`getDerviedStateFromError`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromerror) handler that will track the error automatically.
2. Added an `errorComponent` prop that will accept a custom error component on a per-use basis.
3. After [discussing with design](https://metaboat.slack.com/archives/C02H619CJ8K/p1675091445490549), made the default error a simple warning icon with a tooltip to make it maximally versatile.

alxnddr reviewed (approved):
🏆
